frontend/apps/reader/modules/readerpaging.lua

```diff
@@ -51,7 +51,38 @@ end
 function ReaderPaging:onGesture() end
 
 function ReaderPaging:registerKeyEvents()
-    if Device:hasKeys() then
+    if Device:hasPageUpDownKeys() and Device:hasDPad() then
```

Comment:
As stated I don't like this. It should be either with a specialized property or with a `not Device:hasKeyboard()` added to avoid messing with the d-pad/arrow keys where it's undesirable (i.e., everywhere else).

Reply:
okay, will add it later. just not `hasWeaponSelect`, choose another name ;), seriously choose another name.

Reply:
Always intended as "a concept like" or "along these lines". Weapon select was merely the first thing to come to mind, because it's been used for such a purpose in videogames since primary controls moved over to sticks, freeing them up as extra buttons.

So more generically you get something like `useDpadAsExtraButtons`.

Reply:
sorry but in the end `if Device:hasPageUpDownKeys() and Device:hasDPad() then` is necessary as it in fact includes kindles with keyboards. `useDPad...` implemented elsewhere where appropiate though.

Reply:
But that's the point. We don't want these changes where they aren't necessary. Imho that very much includes Kindles with a keyboard, after all they don't need to give up the precious dpad to start selection, but that's somewhat besides the point.

Reply:
I agree that it isn't precious there, but it would be inconsistent furthermore the proposed change is that much better. Besides, 90% of key interactions (not including page turners) are dPad related, it makes no sense to not to do it proper. As a proud kindle keyboard owner (although not running koreader there) I approve this message.

Reply:
Then you can add the relevant property to those devices. I think it would make the experience significantly worse elsewhere.

Reply:
But i already have, all kindles with dPads have content selection and next/previos chapter. The way the ereader gods intended.

If you want i can do `hasPageUpDown and hasDpad and not isTouch`?

Reply:
But all of those have the problem unless you add `not hasKeyboard` or introduce a special property to introduce it selectively.

Reply:
But being in all non-touch kindles is kind of the point. What other properties? I might as well just flat out do `isKindle` at this point

Reply:
Like I said, just add a new property that you use *as if* it were isKindle but isn't so that it's future -friendly. That's what useDpadetc. is.

Reply:
`readsAZW`? Kindle specific, programmer/user friendly, unlikely to spread to other manufacturers, applies to every single kindle made to date. Happy with that?

Reply:
I apologize that I don't seem to be able to convey why I do not want to see isKindle for something that should be generic, but no, that I'm afraid that is just a very bad name. It doesn't describe what it does at all.

Good code is something you can read and understand a year down the line, and it doesn't unnecessarily restrict you from testing it, nor from deciding that maybe this control scheme is exactly what's desired when you plug in a gamepad.

There are only two options: either you do it with properties that exist and add "not hasKeyboard" or you add a new property with a proper name along the lines of "useDpadAsExtraKeys". The exact name doesn't matter but it must be clear and generic.

Reply:
haven't caught up with the whole thing, so at a glance I'm not sure this necessarily deserves a dedicated cap, so a "not hasKeyboard" sounds fine to me at first blush.

But, if we want a dedicated cap, something along the lines of `hasFiveway` (which is how lab126 refers to the actual hardware). I've heard similar input devices on monitors being called "selection nipple", which, err, no? :D

Reply:
And as a propsotion for a more generic name: `hasSelectionStick`?

Reply:
Oh, looking at the diff, I am *very* much not a fan of the whole `hasPageUpDownKeys` thingy.

It really, really needs to go.

AFAIR (please correct me if I'm wrong), no Kindle devices with a dpad happen to also *not* feature pageturn buttons, so it makes no sense to me.

And if there is, it's probably *one*, so I'd rather handle it as a special case somehow rather than like this.

Reply:
TL;DR: If you need to make a distinction between "hasDPad because we have up/down keys (i.e., everything)" and "hasDPad because it's a Kindle's fivveway stick", then, yeah, just go with a single new, Kindle-specific `hasFiveway` (or similar), and leave Generic alone (except for making it public and default to no, of course).

But that pageupdown thingy is nasty, please, no ;).

Reply:
all kindles with dPads have in fact page turn buttons (pageupdown), I believe Frenzie wanted pageupdown because some pocketbooks have dPads but no page turning buttons so the dpad needs repurposing there. I believe the worst offender is pocketbook 613 with two buttons and a dpad: [ < ] [ [ DPad ] ] [ > ]

Reply:
Oh, right, PB shitting on our parade again ;).

Yeah, I'd rather tackle this the other way around as proposed https://github.com/koreader/koreader/pull/11749#discussion_r1602325433 then ;).

(Assuming I groked what you were going for correctly, of course).

Reply:
as far as I can see
`hasKeys` is anything with at least one button (power buttons don't count)
`hasDPad` that is clear
`hasKeyboard` clear too
`hasFewKeys` is a subset of hasKeys and seems to imply not enough buttons for normal functioning, therefore gymnastics are required to make it work.

Frenzie's valid point is, some devices have `hasDPad` and `hasFewKeys` (therefore `hasKeys` too). so we can't guarantee navigation so let's introduce `hasPageUpDown` to differenciate them.

maybe that could be renamed to `hasPageTurnButtons`

Reply:
> TL;DR: If you need to make a distinction between "hasDPad because we have up/down keys (i.e., everything with a keyboard/gamepad)" and "hasDPad because it's a Kindle's fiveway stick", then, yeah, just go with a single new, Kindle-specific `hasFiveway` (or similar), and leave Generic alone (except for making it public and default to no, of course).

But most notably gamepads might also want to be treated this way, which is where being Kindle specific in wording actively prevents anyone from realizing it's already been implemented. But other than the name that's exactly what I'm saying of course.

Reply:
marking this as resolved then...